### ISSUE TITLE
[RELAY][OP] Relay Operator Sprint

### ISSUE THREAD

**Issue by tqchen** [dmlc/tvm#1799]
Now that the Relay RFC is being merged and we are stabilizing the type inference interface, we should sprint to add new operators to relay to make it on parity with NNVM. 

#1798 shows an example on how to do so for conv2d operator.

## General Steps of Porting
- Implement the TypeRelation function, when necessary 
   - The shapes represented by IndexExpr(symbolic integer)
      - When possible, support symbolic shape inference
      - You can, however, get the integer out from symbolic shape if it is a must, that will require the inference to work on concrete shapes.
   - User reporter->Assign to set the inferred result
   - Use reporter->AssertEQ to assert symbolic integer equivalence
      - It will return false if there is an unsatisfied constraint
- Use tvm::Attrs to replace dmlc::Parameter
- We switch to directly create python wrappers by calling into positional functions so that the operator signature is explicit in python

## General Principles
- Numpy consistency, always consistent with numpy
   - All binary operators broadcast
   - This means we will use add, subtract instead of broadcast_add, broadcast_sub ...
   - elemwise_add version will not be supported for now as we can just use the broadcast version
- Consistent with nnvm when possible
- Fields in Attrs
   - Use concrete types when possible(int, string, bool)
   - If you need None, you can use IndexExpr, which gives you that

## List of Operators to be covered
Generally, we need to cover everything we have so far  https://docs.tvm.ai/nnvm_top.html
Please use this issue to coordinate what you will be working on. As we expect things to move quickly, try to do "fine grained locking" and only claim things that you are working on right now and aim to get things in a few days.

### Level 1: Common Basic Ops
Enough to get MLP
- [x] nn.dense
- [x] nn.relu
- [x] tanh
- [x] sigmoid
- [x] exp 
- [x] log
- [x] sqrt
- [x] add
- [x] subtract
- [x] multiply
- [x] divide
- [x] mod
- [x] nn.batch_flatten
- [x] concatenate
- [x] nn.softmax
- [x] nn.log_softmax
- [x] nn.batch_norm
- [x] nn.dropout
- [x] expand_dims

### Level 2: Convolutions
Enough to get convnet
- [x] nn.conv2d
- [x] nn.conv2d_transpose
- [x] nn.max_pool2d
- [x] nn.avg_pool2d
- [x] nn.global_max_pool2d
- [x] nn.global_avg_pool2d
- [x] nn.pad
- [x] nn.lrn

### Level 3: Additional Math And Transform Operators
- [x] reshape
- [x] copy
- [x] negative
- [x] floor
- [x] ceil
- [x] round
- [x] trunc
- [x] clip
- [x] abs
- [x] leaky_relu
- [x] tranpose 
- [x] split
- [x] squeeze
- [x] take
- [x] full
- [x] zeros
- [x] ones
- [x] transpose
- [x] zeros_like
- [x] ones_like

### Level 4: All broadcast and reduction functions that are not in previous level
- [x] pow
- [x] less
- [x] greater
- [x] less_than
- [x] greater_than
- [x] right_shift
- [x] left_shift 
- [x] maximum
- [x] minimum
- [x] sum
- [x] max
- [x] prod
- [x] argmax, argmin
- [x] strided_slice
- [x] broadcast_to
- [x] where

### Level 5: Vision Operators
- [x] image.resize
- [x] vision.multibox_prior
- [x] vision.nms

### Level 10: Backend Operators
Operators necessary as intermediate stage of optimizations, or gradient, can be influx

**Comment by MarisaKirisame** [comment]
I had 'implemented' some elem wise function which I need for ad (negative, multiplication, division)
('implemented' because it isnt possible to lower code right now)
I can take some operators, and I dont have any particular preference.

**Comment by srkreddy1238** [comment]
I could get through resize operator (will PR once #1798 is merged) to start with and proceeding with transforms.

**Comment by junrushao1994** [comment]
I strongly agree with the numpy consistency, e.g. `nnvm.symbol.flatten` should be renamed.

**Comment by junrushao1994** [comment]
I can work on some shape-related APIs.

**Comment by srkreddy1238** [comment]
@tqchen 
How about relaying an assert on a condition which involve two variables ?
Ref. Transposed convolution where channels should divide groups and both are variables.

**Comment by junrushao1994** [comment]
@srkreddy1238 
I am starting with `expand_dims`, the easiest one

**Comment by tqchen** [comment]
One note about int32 vs int64 when constructing constant was raised by @junrushao1994 @srkreddy1238 . This is an issue we should think about now. int32 will likely cause some regression on large arrays which need to be fixed. I think we should prefer int64 when possible for constants, and let the compiler to automatically detect and downgrade to int32. 

A temporary workaround is always to keep the inferred type consistent with the input  data type, and we can make the switch more easily in one place later

**Comment by junrushao1994** [comment]
Another thing I am concerning is user friendliness.

First, examples provided by Python API docs should be at least runnable by copy-pasting, like PyTorch (https://pytorch.org/docs/stable/tensors.html) or NumPy (https://docs.scipy.org/doc/numpy/reference/generated/numpy.expand_dims.html).

Second, Python API docs should be self-contained, at least those designed for DL practitioners who may not take a good look at the C++ code.

It does not seem to be a big deal for now, but we should put more effort into it in the future.

**Comment by tqchen** [comment]
+1 for API docs friendless, I would recommend we do it now than later. Maybe I am having a bad lead example in the conv2d docs as it was pretty minimum, I will send an updated PR to update that, and let us make sure the new ops are being well documented with examples, especially non-trival ones

**Comment by srkreddy1238** [comment]
Expr like below is not getting simplified !!

```TensorTypeNode(float32, [n, c, int32((float32(100)*2.000000f)), int32((float32(200)*2.000000f))])```

any idea ?

**Comment by tqchen** [comment]
The eager CSE is done among integer expression only so far. For floating points, we still need to call explicitly simplification, or use as_const_int to get out and explicit simplify

**Comment by junrushao1994** [comment]
I am going to grab `transpose`

**Comment by tqchen** [comment]
@junrushao1994 transpose should be in the list, sorry the list was not complete

**Comment by MarisaKirisame** [comment]
I am taking/had taken multiply/divide/mod/relu/tanh/sigmoid/negative.

**Comment by srkreddy1238** [comment]
@MarisaKirisame 
I have covered multiply, divide, mod, tanh, sigmoid, negative already in #1813

**Comment by srkreddy1238** [comment]
To keep all of us on same page #1813  also covers 
multiply, mod, tanh, sigmoid, negative, floor, ceil, trunc, abs, pow, resize, upsampling, batch_flatten, pool2d and global_pool2d

**Comment by junrushao1994** [comment]
`expand_dims` is in #1819 
comparison `greater` `greater_equal` `less` `less_equal` `not_equal` `equal ` is in #1824

**Comment by tqchen** [comment]
take ```right_shift```

**Comment by tmoreau89** [comment]
take `left_shift`

**Comment by MarisaKirisame** [comment]
I had done zeros_like and ones_like. I think I will take zeros and ones too for symmetricity

**Comment by Luo-Liang** [comment]
attempting maximum

**Comment by Mutinifni** [comment]
attempting minimum

**Comment by slyubomirsky** [comment]
Attempting `pad`

**Comment by joshpoll** [comment]
take `clip`

**Comment by cowanmeg** [comment]
take sigmoid

**Comment by merrymercy** [comment]
take softmax

**Comment by slyubomirsky** [comment]
take `full` and `full_like` (the latter appears to be missing from the list even though it's an NNVM op)

**Comment by kevinthesun** [comment]
take mutibox_prior and nms

**Comment by junrushao1994** [comment]
Some random notes: should we, or is there any tools could, do runtime type checking on the relay's Python API side?

**Comment by tqchen** [comment]
@junrushao1994 can you elaborate on what you mean?

**Comment by junrushao1994** [comment]
@tqchen Sorry, I mean sanity check. It is not related to relay's type, just wondering if it is necessary to add some sanity checks to safeguard type safety of cffi.

For example, the signature of some function is ```def f(a: List[int], b: int)```, but python won't check actual type of `a` and `b`. If `f` directly passes `a` and `b` to ctypes, who doesn't seem to do sanity check either. In this case, if `a` and `b` are not of proper types, the code crashes without a informative error message.

**Comment by tqchen** [comment]
I think there is a python parser that does the similar thing but needs to confirm with @jroesch @joshpoll

**Comment by junrushao1994** [comment]
@tqchen Sounds cool!

**Comment by joshpoll** [comment]
You can do this with mypy types, see #1781 for an example. We could add type annotations to the operators, which will provide the sanity checks you want. Mypy is a static analyzer, so in order to get its benefits you need to run it separately or have it integrated into your IDE.

**Comment by srkreddy1238** [comment]
taking conv2d_transpose

**Comment by siju-samuel** [comment]
@tqchen  how to handle multiple outputs? i was trying `dropout` and `batchnorm`, both have multiple outputs.
In `RELAY_REGISTER_OP`, like `set_num_inputs`, do we have something like `set_num_outputs`?
I think currently `Array<Type>& types` will have only (inputs + one output)

**Comment by srkreddy1238** [comment]
multiple outputs is needed for split too.

**Comment by MarisaKirisame** [comment]
@siju-samuel @srkreddy1238 multiple output is possible only by wrapping all of them in a tuple type

**Comment by slyubomirsky** [comment]
I'll take `batch_norm` and `dropout` as well, to finish out level 1

**Comment by siju-samuel** [comment]
i started with reduce ops, you can do with batch_norm & dropout

**Comment by zhiics** [comment]
attempting where

**Comment by srkreddy1238** [comment]
Attempting split to conclude Level 3

**Comment by anijain2305** [comment]
Will attempt prod

**Comment by jroesch** [comment]
Thanks to everyone for the hard work on getting 99% of the way there. I'm making a push to now add the compute and scheduling behavior for all of these operators which should enable users to use Relay for end-to-end inference tasks, enable new frontends and more. If you you would be interested in helping read more here: https://github.com/dmlc/tvm/issues/2051.

**Comment by yzhliu** [comment]
@tqchen I believe all listed operators have been implemented, could you double check?

**Comment by tqchen** [comment]
Thanks to everyone for the hard work, this issue is closed as most ops are in, we will follow up in #2051
